src/device/wg_controller.rs

```diff
@@ -642,14 +642,21 @@ async fn build_requested_meshnet_peers_list<
         // Select actual endpoint
         let (selected_remote_endpoint, selected_local_endpoint) = select_endpoint_for_peer(
             public_key,
-            &actual_peer.cloned(),
-            &time_since_last_rx_or_handshake,
+            actual_peer,
+            time_since_last_rx_or_handshake.as_ref(),
             peer_state,
-            &checked_endpoint.cloned(),
-            &proxy_endpoint.cloned(),
-            &upgrade_request_endpoint,
-        )
-        .await?;
+            checked_endpoint,
+            proxy_endpoint,
+            upgrade_request_endpoint.as_ref(),
+        );
+
+        // If checked endpoint is consumed, we need to remove it from CPC
+        if selected_remote_endpoint == checked_endpoint.map(|ep| ep.remote_endpoint) {
```

Comment:
maybe it would be better to call `consume_verified_endpoint()` in the place where we in-fact advertise it to the other side via the `request_upgrade()` call?

I am afraid, that there are several steps between here and the endpoint being actually requested for upgrade, that some condition may override it, and then we would be consuming it incorrectly.